### PR TITLE
bazel: make work on nixos :sparkle: 

### PR DESCRIPTION
**Pull request by Strum355**
Some sprinkling of special sauce for the NixOS (and nix shell) users in the house. Much care has been taken so that there should be no interference with anyone running bazel on a non-nixos machine, and that it also works as (mostly) expected for nix shell users on macos.

## Test plan

Ran main dry-run, all green, so it shouldnt interfere with anyone on a non-nixos machine

### CODE REVIEW

File: WORKSPACE [hunk 1]
```diff
@@ -74,19 +74,50 @@ http_archive(
     urls = ["https://github.com/bazelbuild/rules_rust/releases/download/0.19.0/rules_rust-v0.19.0.tar.gz"],
 )
 
+http_archive(
```
Comment (jhchabran) [inline WORKSPACE:77]: Can we move this into `dev/nix.bzl` ? 

You can basically take all of this code, and wrap it under a macro: 

```
# dev/nix.bzl 
def nix_deps(): 
  http_archive(...) 
```

```
# WORKSPACE
load("//dev:nix.bzl", "nix_deps") 
nix_deps() 
``` 

This way we avoid lengthening this file further 😊

**Review by jhchabran — APPROVED** [review]
Approving, because the little suggestion I made is pretty straightforward and it would be okay to merge as is any way, as it's mostly cosmetic :)

File: WORKSPACE (hunk 1 above)
Comment (Strum355) [reply WORKSPACE:77]: Done as best I could [here](https://github.com/sourcegraph/sourcegraph/pull/51154/commits/c92076f72dab9f2e3bc4e54ed3b76e63f7c298fb), it seems like `rules_nixpkgs_dependencies` must be called before any of the other symbols are loaded, so some stuff remains in WORKSPACE

**Comment by Strum355** [comment]
Will need to rebase this now that https://github.com/sourcegraph/sourcegraph/pull/49884 landed, job for tomorrow

**Pull request merged**